### PR TITLE
[Kafka output] removed regex validation to allow dynamic topic

### PR DESCRIPTION
**Pull request by juliaElastic**
## Proposed commit message

Removed regex validation to allow dynamic topic.

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## How to test this PR locally

- build beats, copy over agentbeat to the elastic-agent build
``` 
# build beats agentbeat
cd ~/beats/x-pack/agentbeat 
SNAPSHOT=true PLATFORMS="darwin/amd64" mage package

# build agent
cd ~/elastic-agent
DEV=true EXTERNAL=false SNAPSHOT=true PLATFORMS=darwin/amd64 PACKAGES=tar.gz mage -v package

# copy agentbeat
cp ~/beats/x-pack/agentbeat/build/golang-crossbuild/agentbeat-darwin-amd64 ~/elastic-agent/build/distributions/elastic-agent-8.16.0-SNAPSHOT-darwin-x86_64/data/elastic-agent-3f22cc/components/agentbeat
```

To test with a real kafka server, follow this guide: https://hevodata.com/learn/install-kafka-on-mac/#Step_2_Download_Install_Kafka_on_Mac_Manually_or_via_HomeBrew
- downloaded and started zookeeper and kafka server
- created a topic `system.cpu` and started the script to consume messages
```
cd ~/Downloads/kafka_2.13-3.8.0
./bin/zookeeper-server-start.sh config/zookeeper.properties
./bin/kafka-server-start.sh config/server.properties
./bin/kafka-topics.sh --create --bootstrap-server localhost:9092 --replication-factor 1 --partitions 1 --topic system.cpu
./bin/kafka-console-consumer.sh --bootstrap-server localhost:9092 --topic system.cpu
```
- added kafka output by entering host `localhost:9092` and topic `%{[data_stream.dataset]}`
- after enrolling the agent, see messages showing up in the kafka consumer
-  in agent logs, expect no errors

image
image
image

## Related issues

- Closes https://github.com/elastic/ingest-dev/issues/3618

### CODE REVIEW

**Comment by faec** [comment]
This check was added intentionally to block dynamic topics in Agent, since they can't be carried over in the OTel migration. Has this decision changed? The code looks fine but last I knew there was a business decision not to support it.

**Comment by juliaElastic** [comment]
> This check was added intentionally to block dynamic topics in Agent, since they can't be carried over in the OTel migration. Has this decision changed? The code looks fine but last I knew there was a business decision not to support it.

We decided to add it back dynamic topics support, see https://github.com/elastic/ingest-dev/issues/3407
Perhaps we could make the validation stricter to only allow a specific field `%{[FIELD]}`, not any format.

File: libbeat/outputs/kafka/kafka.go
Comment (juliaElastic) [inline libbeat/outputs/kafka/kafka.go:1]: trying to undo the changes added here: https://github.com/elastic/beats/pull/37902/files#diff-d8e7a8c5739c80fb30e5c2f22ff87b2c8d476d8e9b71c222f2bcd6b4a6d0c3d6
though the dynamic topic resolution still doesn't seem to work

Comment (juliaElastic) [reply libbeat/outputs/kafka/kafka.go:1]: @cmacknz Can I get some help/review on this? I'm not sure why the dynamic topic is not being resolved in agent. Tried to add debug logs locally, but I'm not seeing the `buildTopicSelector` function being triggered.

Comment (cmacknz) [reply libbeat/outputs/kafka/kafka.go:1]: Hmm, I see some things to adjust but nothing that would cause that function not to be hit at all?

How are you testing this? You now need to build `x-pack/agentbeat` and copy over the resulting `agentbeat` binary, so make sure you are doing that and not copying one of the beat binaries directly.

File: libbeat/outputs/kafka/config.go [hunk 1]
```diff
@@ -112,7 +112,8 @@ var compressionModes = map[string]sarama.CompressionCodec{
 // validTopicRegExp is used to validate the topic contains only valid characters
 // when running under Elastic-Agent. The regexp is taken from:
 // https://github.com/apache/kafka/blob/a126e3a622f2b7142f3543b9dbee54b6412ba9d8/clients/src/main/java/org/apache/kafka/common/internals/Topic.java#L33
-var validTopicRegExp = regexp.MustCompile("^[a-zA-Z0-9._-]+$")
+// also allowing a format string that contains a field name %{[FIELD]}
+var validTopicRegExp = regexp.MustCompile(`^(?:%\{\[)?[a-zA-Z0-9._-]+(?:\]\})?$`)
```
Comment (cmacknz) [inline libbeat/outputs/kafka/config.go:116]: I think we can probably get rid of this, it was meant to catch people including the topic selector syntax before it hit Kafka and matched the valid kafka topic syntax exactly.

If we were to keep the valid topic regex, the place to put it would be after the topic substitution has already happened.

File: libbeat/outputs/kafka/kafka.go
Comment (juliaElastic) [reply libbeat/outputs/kafka/kafka.go:1]: I didn't copy anything, I thought the `EXTERNAL=false` flag should build and use the local `beats` repo. I'll try copying it.

Comment (juliaElastic) [reply libbeat/outputs/kafka/kafka.go:1]: okay I could build beats, copied over to agent and now I can see the dynamic topic resolution works as expected, I'll update the description with screenshots

File: libbeat/outputs/kafka/config.go (hunk 1 above)
Comment (juliaElastic) [reply libbeat/outputs/kafka/config.go]: removed the validation

File: CHANGELOG.next.asciidoc [hunk 2]
```diff
@@ -191,6 +191,7 @@ https://github.com/elastic/beats/compare/v8.8.1\...main[Check the HEAD diff]
 - elasticsearch output now supports `idle_connection_timeout`. {issue}35616[35615] {pull}36843[36843]
 - Enable early event encoding in the Elasticsearch output, improving cpu and memory use {pull}38572[38572]
 - The environment variable `BEATS_ADD_CLOUD_METADATA_PROVIDERS` overrides configured/default `add_cloud_metadata` providers {pull}38669[38669]
+- Kafka output allows dynamic topic in `topic` field {pull}40415[40415]
```
Comment (belimawr) [inline CHANGELOG.next.asciidoc:194]: I believe it would be good to include that this is only under Elastic-Agent because standalone Beats have always supported this.
```suggestion
- When running under Elastic-Agent Kafka output allows dynamic topic in `topic` field {pull}40415[40415]
```

File: libbeat/outputs/kafka/kafka.go
Comment (cmacknz) [reply libbeat/outputs/kafka/kafka.go:1]: Probably `EXTERNAL=false` isn't updated to be aware of agentbeat yet.

File: libbeat/outputs/kafka/config.go (hunk 1 above)
Comment (cmacknz) [reply libbeat/outputs/kafka/config.go]: We can probably just remove it entirely, this validation also exists in the Kafka API when we go to create the topic.

We were trying to catch the problem earlier with this and make things more obvious if users pasted examples that used to work when we disabled it.

If we did keep it we'd put in the block below, which is at the point where he message is published and so is per event and doesn't buy us anything over what the Kafka API is going to give us.

https://github.com/elastic/beats/blob/d7ae68c7aa40aa27deff6b90508bb8f8664a6efa/libbeat/outputs/kafka/client.go#L215-L221

Comment (juliaElastic) [reply libbeat/outputs/kafka/config.go]: Thanks, I'll leave it removed then.

File: libbeat/outputs/kafka/kafka.go
Comment (juliaElastic) [reply libbeat/outputs/kafka/kafka.go:1]: I think I missed the part where the beats are not rebuilt if it was built already, and the step to copy it over to agent.

**Comment by juliaElastic** [comment]
@belimawr @cmacknz Are you ready to approve? I would like to get this in before the sprint ends this week.

**Review by belimawr — APPROVED** [review]
